### PR TITLE
BUG fixes tuple agg  issue 18079

### LINKED ISSUE #18079: DataFrameGroupBy.aggregate can not work with `tuple` as an argument
The following code raises `ValueError` 
```python
grouped_df = df.groupby(group_by_attributes, as_index=False).aggregate(tuple)
```
Here is a more replicatable version:

```python
import pandas as pd
import numpy as np
df = pd.DataFrame(np.random.randn(100, 3), columns=list('ABC'))
grouped_df = df.groupby(['A', 'B'], as_index=False).aggregate(tuple)
```
#### Problem description

The statement above does not work because `tuple` is not a function. It throws: 
`ValueError: no results`

#### Workaround
use the following `groupby` statement instead
```python
grouped_df = df.groupby(['A', 'B'], as_index=False).aggregate(lambda x: tuple(x))
```
This was issued as a result of the following discussion: https://github.com/PyCQA/pylint/issues/1709#issuecomment-341095096

#### Expected Output
Should be able to work without raising a `ValueError` 
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-97-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.20.3
pytest: None
pip: 8.1.2
setuptools: 28.2.0
Cython: 0.24.1
numpy: 1.13.3
scipy: 0.18.1
xarray: None
IPython: 5.4.1
sphinx: 1.4.8
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 1.5.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by bobhaffner**
- [x] closes #18079
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.1.txt [hunk 1]
```diff
@@ -62,6 +62,7 @@ Bug Fixes
 - Bug in ``pd.Series.rolling.skew()`` and ``rolling.kurt()`` with all equal values has floating issue (:issue:`18044`)
 - Bug in ``pd.DataFrameGroupBy.count()`` when counting over a datetimelike column (:issue:`13393`)
 - Bug in ``pd.concat`` when empty and non-empty DataFrames or Series are concatenated (:issue:`18178` :issue:`18187`)
+- Bug in :class:`NDFrameGroupBy` fixes ValueError: no results error when grouping by a single column and aggregating with a tuple (:issue:`18079`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.1.txt]: Maybe just say something like "Bug when grouping by a single column and aggregating with a class like ``list`` or ``tuple``" . I don't recall if NDFRameGroupBy is in the public API.

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -2718,6 +2718,22 @@ def test_empty_dataframe_groupby(self):
...
+        df = pd.DataFrame({'A': [1, 1, 1, 3, 3, 3],
+                           'B': [1, 1, 1, 4, 4, 4], 'C': [1, 1, 1, 3, 4, 4]})
+
+        result = df.groupby(['A', 'B']).aggregate(tuple)
+        expected = pd.DataFrame({'C': {(1, 1): (1, 1, 1), (3, 4): (3, 4, 4)}})
+        expected.index.names = ['A', 'B']
+        assert_frame_equal(result, expected)
+
+        result = df.groupby('A')['C'].aggregate(tuple)
+        expected = pd.Series([(1, 1, 1), (3, 4, 4)], index=[1, 3])
```
Comment (TomAugspurger) [inline pandas/tests/groupby/test_groupby.py]: Should be able to pass `name='C'` here.

**Comment by bobhaffner** [comment]
@TomAugspurger's  comment on whats new prompted me to add a test for `list` aggregation only to discover a new issue.   

This works
df.groupby('A')['C'].aggregate(list)

This doesn't.  
df.groupby(['A', 'B']).aggregate(list)

Should I submit a separate issue for this or investigate and fix this new error as part of this PR?

<details>
  <summary>Click to see full traceback</summary>
  ---------------------------------------------------------------------------

Traceback (most recent call last):
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2267, in agg_series
    return self._aggregate_series_fast(obj, func)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2287, in _aggregate_series_fast
    result, counts = grouper.get_result()
  File "pandas/_libs/src/reduce.pyx", line 406, in pandas._libs.lib.SeriesGrouper.get_result
    raise
  File "pandas/_libs/src/reduce.pyx", line 394, in pandas._libs.lib.SeriesGrouper.get_result
    result = _get_result_array(res,
  File "pandas/_libs/src/reduce.pyx", line 15, in pandas._libs.lib._get_result_array
    raise ValueError('function does not reduce')
ValueError: function does not reduce

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 4192, in aggregate
    return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 3634, in aggregate
    return self._python_agg_general(arg, *args, **kwargs)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 853, in _python_agg_general
    result, counts = self.grouper.agg_series(obj, f)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2269, in agg_series
    return self._aggregate_series_pure_python(obj, func)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2304, in _aggregate_series_pure_python
    raise ValueError('Function does not reduce')
ValueError: Function does not reduce

</details>

File: pandas/tests/groupby/test_groupby.py [hunk 3]
```diff
@@ -2718,6 +2718,21 @@ def test_empty_dataframe_groupby(self):
 
         assert_frame_equal(result, expected)
 
+    def test_tuple_aggregation(self):
+        # Issue #18079
+        df = pd.DataFrame({'A': [1, 1, 1, 3, 3, 3],
+                           'B': [1, 1, 1, 4, 4, 4], 'C': [1, 1, 1, 3, 4, 4]})
+
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: move to test_aggregate

can you parametrize with list-likes here (e.g. tuple/list,np.array,Series)

some of these might have different output (Series) but let's see.

you may have to have a tests with the parameterized ones and another one with 'other' things, e.g. 

```
def f(x):
    return tuple(x)
```

which work now.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_groupby.py (hunk 3 above)
Comment (bobhaffner) [reply pandas/tests/groupby/test_groupby.py]: Thanks for the feedback, @jreback!  I'm hoping to tackle the list bug and these additional tests later this week.

**Comment by bobhaffner** [comment]
- Moved tests to test_aggregate.py

- Parametrize the tests (one for DataFrames and one for Series))

- I fixed the list issue I mentioned above by removing the list condition in the following if

https://github.com/pandas-dev/pandas/blob/4fce7846be56e12999fe8758abb2ea2f2794259d/pandas/core/groupby.py#L2303

I was a little leery doing this, but all tests passed.

Note: Both np.array and pd.Series still error out with both scenarios (grouping by one column vs multiple columns), but with different error descriptions.  That said, is it possible to proceed with this PR since both tuple and list now behave the same way with single and multiple column groupbys?
```
df = pd.DataFrame({'A' : [1, 1, 3], 'B' :  [1, 1, 4], 'C' :  [1, 3, 4]})

df.groupby('A')['C'].aggregate(pd.Series) # np.array
>> Exception: Must produce aggregated value

df.groupby(['A', 'B']).aggregate(pd.Series) # np.array
>> ValueError: Function does not reduce
```

This might close https://github.com/pandas-dev/pandas/issues/4293 as well

File: doc/source/whatsnew/v0.21.1.txt [hunk 4]
```diff
@@ -103,7 +103,7 @@ Groupby/Resample/Rolling
 - Bug in ``DataFrame.resample(...).apply(...)`` when there is a callable that returns different columns (:issue:`15169`)
 - Bug in ``DataFrame.resample(...)`` when there is a time change (DST) and resampling frequecy is 12h or higher (:issue:`15549`)
 - Bug in ``pd.DataFrameGroupBy.count()`` when counting over a datetimelike column (:issue:`13393`)
--
+- Bug when grouping by a single column and aggregating with a class like`list` or `tuple` (:issue:`18079`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: use double-backticks around ``list`` and ``tuple``

Comment (jreback) [reply doc/source/whatsnew/v0.21.1.txt]: move this to 0.22 as well. (its not that I don't regard this as a bug fix, but this is something touching an important piece of code and want this to live in master for a bit).

**Review by jreback — CHANGES_REQUESTED** [review]
small changes, otherwise lgtm. ping on green.

**Comment by bobhaffner** [comment]
Thanks @jreback!

**Comment by jreback** [comment]
one more rebase :>

**Comment by bobhaffner** [comment]
Hi @jreback I had conflicts (whatsnew) with the rebase, but I believe Its corrected now.  That said, I'm puzzled by the pep8speaks comment regarding `test_agg.py` though as I didn't modify that file.  My apologies for the confusion.

File: doc/source/whatsnew/v0.21.1.txt.orig [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+Documentation Changes
+~~~~~~~~~~~~~~~~~~~~~
+
+-
+-
+-
+
+.. _whatsnew_0211.bug_fixes:
+
+Bug Fixes
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt.orig]: remove this

File: grp.patch [hunk 6]
```diff
@@ -0,0 +1,13 @@
+--- a/pandas/core/groupby.py
++++ b/pandas/core/groupby.py
+@@ -3022,7 +3022,9 @@ class SeriesGroupBy(GroupBy):
+         if isinstance(func_or_funcs, compat.string_types):
+             return getattr(self, func_or_funcs)(*args, **kwargs)
+ 
+-        if hasattr(func_or_funcs, '__iter__'):
```
Comment (jreback) [inline grp.patch]: remove this

File: test_agg.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import pandas as pd
+import numpy as np
+
+def f(x):
+   return list(x)
+
+#df = pd.DataFrame({'A' : [1, 1, 3], 'B' :  [1, 2, 4]})
+#result = df.groupby('A').aggregate(f)
```
Comment (jreback) [inline test_agg.py]: remove this

**Review by jreback — CHANGES_REQUESTED** [review]
looks like some extra files got added during the rebase. pls remove them. and push again.

**Comment by bobhaffner** [comment]
Yikes, my mistake.  thanks Jeff.

File: grp_test.patch [hunk 8]
```diff
@@ -0,0 +1,15 @@
+--- a/pandas/tests/groupby/test_groupby.py
++++ b/pandas/tests/groupby/test_groupby.py
+@@ -2725,3 +2725,12 @@ def _check_groupby(df, result, keys, field, f=lambda x: x.sum()):
+     expected = f(df.groupby(tups)[field])
+     for k, v in compat.iteritems(expected):
+         assert (result[k] == v)
```
Comment (jreback) [inline grp_test.patch]: patch needs removal

File: doc/source/whatsnew/v0.21.1.txt [hunk 9]
```diff
@@ -147,4 +147,4 @@ Other
 ^^^^^
 
 -
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: can you revert this file

**Review by jreback — CHANGES_REQUESTED** [review]
couple of more items to remove

File: doc/source/whatsnew/v0.21.1.txt (hunk 9 above)
Comment (bobhaffner) [reply doc/source/whatsnew/v0.21.1.txt]: HI Jeff, I'm afraid my lack of git experience is showing as this file has been a bit of a thorn in my side.   I'm concerned that my attempts to fix my mess will just result in wasting more of your time.  Would you mind sharing the syntax necessary to revert (or reset?) this file to the proper commit?

Comment (jreback) [reply doc/source/whatsnew/v0.21.1.txt]: sure ill fix this upl

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
should be good to go. ping on green.

**Comment by bobhaffner** [comment]
Thanks Jeff!  You're a lifesaver!

**Comment by jreback** [comment]
thanks @bobhaffner

**Comment by bobhaffner** [comment]
Many thanks for the help and the patience, @jreback .   Same to you @TomAugspurger

**Pull request merged**